sentinelhub/download/session.py

```diff
@@ -126,3 +134,125 @@ def _fetch_token(self, request: DownloadRequest) -> JsonDict:
             return oauth_session.fetch_token(
                 token_url=request.url, client_id=self.config.sh_client_id, client_secret=self.config.sh_client_secret
             )
+
+
+_DEFAULT_SESSION_MEMORY_NAME = "sh-session-token"
+_NULL_MEMORY_VALUE = b"\x00"
+
+
+class SessionSharingThread(Thread):
+    """A thread for sharing a token from `SentinelHubSession` object in a shared memory object that can be accessed by
+    other Python processes during multiprocessing parallelization.
+
+    How to use it:
+
+    .. code-block:: python
+
+        thread = SessionSharingThread(session)
+        thread.start()
+
+        # Run a parallelization process here
+        # Use collect_shared_session() to retrieve the session with other processes
+
+        thread.stop()
+    """
+
+    _EXTRA_MEMORY_BYTES = 100
+
+    def __init__(
+        self, session: SentinelHubSession, *args: Any, memory_name: str = _DEFAULT_SESSION_MEMORY_NAME, **kwargs: Any
+    ):
+        """
+        :param session: A Sentinel Hub session to be used for sharing its authentication token.
+        :param args: Positional arguments to be propagated to `threading.Thread` parent class.
+        :param memory_name: A unique name for the requested shared memory block.
+        :param kwargs: Keyword arguments to be propagated to `threading.Thread` parent class.
+        """
+        super().__init__(*args, **kwargs)
+
+        self.session = session
+        self.memory_name = memory_name
+
+        if self.session.refresh_before_expiry is None:
+            raise ValueError(f"Given instance of {self.session.__class__.__name__} must be self-refreshing")
+        self._refresh_time = self.session.refresh_before_expiry
+
+        self._stop_event = Event()
+        self._is_memory_shared_event = Event()
+
+    def start(self) -> None:
+        """Start running the thread.
+
+        After starting the thread it also waits for the token to be shared. This way no other process would try to
+        access the memory before it even exists."""
+        super().start()
+        self._is_memory_shared_event.wait()
+
+    def run(self) -> None:
+        """A running thread is running an infinite loop of sharing a token and waiting for token to expire. The loop
+        ends only when the thread is stopped."""
+        self._stop_event.clear()
+
+        while not self._stop_event.is_set():
+            token = self.session.token
+            self._share_token(token)
+
+            sleep_until_refresh_time = token["expires_at"] - time.time() - self._refresh_time
```

Comment:
wouldn't this refresh the token JUST as it becomes expired?
does that mean there is a small possibility of the token being grabbed after it expired but before it is refreshed? Can we put a lock on the shared memory while refreshing (and a bit of a buffer to the time where that happens so we're not a microsecond too late)?

Reply:
`token["expires_at"] - time.time()` is the time until the token expires. With additional ` - self._refresh_time` we get the time until session will start refreshing token. So as long a `self._refresh_time` is not `0` (by default it is `120`), there will be still enough time to use the token for download.

Note that the shared memory is updated with a single command. This way I'm sure shared memory implementation makes sure that read and write operations cannot happen at the same time.

Reply:
ah, i see, we're waiting for the window to start, should be a good enough buffer yeah.

probably not even god knows what python considers to be an atomic operation :joy:  but the buffer is large enough so that I dont think we need to worry